### PR TITLE
fix(backend/copilot): re-prompt on thinking-only finish; route storage-limit through DB-manager

### PR DESCRIPTION
**Pull request by majdyz**
## Why

Two production fixes surfaced from John Ababseh's dev testing on 2026-05-01 (Discord thread `1499923303609925793`):

- **Issue #5** — chat session `c93dc51f-bb38-4427-975a-6dc033358689` finished after multiple minutes of work and showed only `(Done — no further commentary.)` Langfuse trace `7d1a674eb7c84ffb5a4b34875306eea9` shows the model wrote the entire restaurant-list answer **inside an extended-thinking `ThinkingBlock`** (931 completion tokens, $0.50 spend) and ended the turn with empty `content: []`. Our existing thinking-only guard immediately stamped the placeholder, so the user never saw the actual answer the model already generated.
- **Issue #2** — every image-generation request (`AIImageCustomizerBlock` / `AIImageGeneratorBlock`) on dev failed with `prisma.errors.ClientNotConnectedError: Client is not connected to the query engine`. Regression from #12780 (tier-based workspace file storage limits): the new pre-write quota check at `util/workspace.py:225` called `get_workspace_total_size` directly from `backend.data.workspace`, which is a Prisma read. The copilot-executor process doesn't connect Prisma — it RPCs into `database-manager` for everything else — so every `manager.write_file()` from a tool blew up.

## What

- **Issue 5** — layered fallback for thinking-only final turns:
  1. Adapter sets `pending_thinking_only_reprompt` and defers placeholder/StreamFinish.
  2. Driver re-enters the SDK loop and fires one synthetic `client.query("Please write a brief user-facing summary of what you found...")`.
  3. If the re-prompt also returns thinking-only, promote the most recent `ThinkingBlock` content to a visible `TextDelta`.
  4. Only when thinking is also empty, emit the original `(Done — no further commentary.)` placeholder.
  Bounded to **one** re-prompt per turn so the worst case is ~one extra LLM call.

- **Issue 2** — route the storage-limit pre-check through the existing `workspace_db()` accessor and expose `get_workspace_total_size` on `DatabaseManager` so the copilot-executor RPCs into database-manager (where Prisma is connected), the same path other workspace queries on this codepath use.

## How

`backend/copilot/sdk/response_adapter.py`
- New `pending_thinking_only_reprompt`, `thinking_only_reprompted`, `_last_thinking_content` fields on `SDKResponseAdapter`.
- Capture latest `block.thinking` when streaming reasoning so the second-tier promote-fallback has content.
- ResultMessage thinking-only branch — first hit defers; second hit prefers `_last_thinking_content`, falls back to placeholder.

`backend/copilot/sdk/service.py`
- Wrap the `async for sdk_msg in _iter_sdk_messages(client):` block in a `while True:` retry loop. After the inner loop ends, check `pending_thinking_only_reprompt` — if set and not yet retried, fire `client.query(_THINKING_ONLY_REPROMPT, ...)` and re-enter; else break. Most of the diff is +4-space indentation churn.
- Module-level `_THINKING_ONLY_REPROMPT` constant for the re-prompt copy.

`backend/data/db_manager.py`
- Import `get_workspace_total_size` and expose it via `_(...)` so it becomes an RPC on `DatabaseManager` and the corresponding async client.

`backend/util/workspace.py`
- Drop the direct `get_workspace_total_size` import; call `workspace_db().get_workspace_total_size(self.workspace_id)` instead.

`backend/util/workspace_test.py`, `backend/copilot/sdk/response_adapter_test.py`
- Existing thinking-only test split into three: defer-on-first-pass, promote-thinking-on-second-pass, fallback-to-placeholder-when-no-thinking.
- Updated `test_flush_unresolved_at_result_message` to expect deferral instead of immediate placeholder.
- New `test_write_file_storage_check_routes_through_workspace_db_accessor` proving the storage-limit pre-check goes through the accessor (would have caught Issue 2).

## Test plan

- [x] `poetry run pytest backend/copilot/sdk/response_adapter_test.py backend/util/workspace_test.py` — 67 pass
- [x] `poetry run ruff check` on changed files — clean
- [x] `poetry run black` / `poetry run isort` on changed files — clean
- [x] `/pr-test --fix` against dev preview to exercise the re-prompt + image-write paths end-to-end
- [x] `/pr-polish` until merge-ready

## Related

- Regression introduced by #12780 (tier-based workspace file storage limits)

### CODE REVIEW

**Comment by majdyz** [comment]
Addressed bot feedback in 99b0aff4d2 (and 2498c6b9a4 from the earlier round):

| # | Source | Finding | Resolution |
|---|--------|---------|------------|
| 1 | sentry | `_any_tool_results_seen` reset to False before re-prompt → second-pass guard never fires | Fixed in `2498c6b9a4` — keep it sticky across the round |
| 2 | sentry | Synthetic re-prompt not added to TranscriptBuilder → divergence with GCS-restored CLI session | Resolved by #6 below — we now strip the re-prompt from the CLI JSONL too, so neither side has it (consistent) |
| 3 | coderabbit | Tests only seed `_last_thinking_content` directly, miss the cross-tool-boundary case | Added `test_tool_result_clears_stale_thinking_so_fallback_does_not_leak_pre_tool_thinking` |
| 4 | coderabbit | `_last_thinking_content` not cleared on tool_result → stale pre-tool reasoning leaks into fallback | Reset alongside `_text_since_last_tool_result` in the tool_result branch |
| 5 | coderabbit | `thinking_only_reprompted` lives on the adapter; transient retries rebuild the adapter and the per-turn cap resets | Promoted to `_RetryState` and propagated to the new adapter on rebuild |
| 6 | coderabbit | `client.query(reprompt)` persists into the CLI JSONL → leaks into `--resume` as a phantom user turn | New `_strip_synthetic_reprompt_from_cli_jsonl` filter applied before `upload_transcript`, with unit coverage for list/string content, image-block preservation, empty input, malformed lines |

All 150 tests across `response_adapter_test.py`, `service_test.py`, `workspace_test.py` pass.

**Comment by majdyz** [comment]
## E2E Test Report

Native dev stack (`poetry run app` + `pnpm dev`, docker only for postgres / redis-cluster / rabbitmq / supabase). Subscription-mode Claude Code auth via OAuth token from macOS keychain.

### Issue 5 — copilot thinking-only fallback
**Repro:** "What are the best restaurants in London? use web search" (extended_thinking)

- Session `f796a38e-643b-44ee-baf4-646e305845a0`
- `web_search` tool returned 3284 bytes; final ResultMessage success, num_turns=2, output=497 tokens
- Persisted message roles: `[user, reasoning, assistant, tool, reasoning, assistant]`
- Final assistant message: 1107 chars of structured London restaurant recommendations + Michelin Guide link
- NO `(Done — no further commentary.)` placeholder

In this run the model produced real text alongside thinking, so the new re-prompt path was not triggered — the layered fallback (re-prompt → promote-thinking → placeholder) is in place but the happy path didn't need it. Verdict: **PASS**.

### Issue 2 — workspace storage-limit pre-check on copilot_executor
**Repro:** Asked the copilot to use `write_workspace_file` to save a 75-byte file — the same `WorkspaceManager.write_file` path the image-gen blocks call.

- Session `91b025d4-f1d8-4c65-906c-85c0b093f062`
- `write_workspace_file` tool invoked **from copilot_executor process** (no direct Prisma client)
- Pre-check `workspace_db().get_workspace_total_size()` ran via DB-manager RPC
- File created: `072e2848-f084-4396-be54-a52be4ee1203 ... size=75 bytes`
- Result: `File saved successfully to your workspace`, exec 7.34s
- NO `ClientNotConnectedError`, NO `prisma.errors.*`

Verdict: **PASS**. The original `AIImageGeneratorBlock` repro could not be exercised (Replicate API key empty in test env), but `write_workspace_file` hits the exact same storage-limit pre-check, so the fix is verified end-to-end.

### Out of scope
- This PR's diff is ~7000 lines / 130+ files (settings v2, onboarding, billing, MCP UI, etc.); only the two production bugs called out in the title were manually exercised here.

### Safe to merge from a runtime perspective: yes.

### Screenshots

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 1]
```diff
@@ -2525,542 +2595,573 @@ async def _run_stream_attempt(
...
+            ):
+                state.adapter.pending_thinking_only_reprompt = False
+                state.thinking_only_reprompted = True
+                state.adapter.thinking_only_reprompted = True
+                # Keep ``_any_tool_results_seen`` True — the guard at
+                # ``ResultMessage`` time still needs it to fire on the
+                # re-prompt round if the model returns thinking-only
+                # again.  Only reset the per-tool-result text-tracker;
+                # if the re-prompt emits a TextBlock it'll set this to
+                # True and the guard won't fire.
```
Comment (majdyz) [inline autogpt_platform/backend/backend/copilot/sdk/service.py]: 🤖 🟡 **Nice to Have**: The 6-line narrative comment here explains *why* `_any_tool_results_seen` is left alone and *why* `_text_since_last_tool_result` is reset — that's reasoning, not invariants. Per AGENTS.md / minimal-comments rule, compress to one line:

```python
# Re-prompt round must still trip the placeholder guard if model returns thinking-only again.
state.adapter._text_since_last_tool_result = False
```

The exact mechanic (why `_any_tool_results_seen` stays True) is now self-evident from the field name + the guard logic in response_adapter.py:444–490; comments here just risk drifting out of sync.

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 2]
```diff
@@ -2525,542 +2534,567 @@ async def _run_stream_attempt(
...
+                if acc.stream_completed:
+                    break
+            if (
+                state.adapter.pending_thinking_only_reprompt
+                and not state.adapter.thinking_only_reprompted
+            ):
+                state.adapter.pending_thinking_only_reprompt = False
+                state.adapter.thinking_only_reprompted = True
+                state.adapter._text_since_last_tool_result = False
+                state.adapter._any_tool_results_seen = False
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py]: 🤖 Already addressed in [2498c6b9a](https://github.com/Significant-Gravitas/AutoGPT/commit/2498c6b9a4): the driver no longer resets `_any_tool_results_seen` between rounds — it now only resets `_text_since_last_tool_result` (see service.py:3145). Verified by `test_result_success_thinking_only_two_rounds_with_driver_reset_emits_fallback` in response_adapter_test.py which exercises exactly this flow (round 1 thinking-only → driver reset → round 2 thinking-only → fallback fires).

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 3]
```diff
@@ -2525,542 +2534,567 @@ async def _run_stream_attempt(
...
+                acc.stream_completed = False
+                logger.info(
+                    "%s Re-prompting model for closing summary "
+                    "after thinking-only final turn",
+                    ctx.log_prefix,
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py]: 🤖 Not applicable — this is intentional design. The synthetic re-prompt is **stripped** from the persisted CLI JSONL (`_strip_synthetic_reprompt_from_cli_jsonl` at service.py:213) precisely so it does NOT leak into `--resume` history on subsequent turns. Adding it to `TranscriptBuilder` would CAUSE the corruption flagged here, not prevent it: the GCS-restored transcript would contain a synthetic user turn the model never actually received from the user. The strip+skip pair keeps both the local CLI session file and the uploaded GCS transcript consistent — only real user input is preserved.

Verified by `TestStripSyntheticReprompt` (service_test.py:1234) covering match/exact-string/string-content variants.

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 4]
```diff
@@ -2525,542 +2595,573 @@ async def _run_stream_attempt(
...
+                    _msgs_since_flush = 0
 
-            if acc.stream_completed:
-                break
+                if acc.stream_completed:
+                    break
+            if (
+                state.adapter.pending_thinking_only_reprompt
+                and not state.thinking_only_reprompted
+            ):
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py]: 🤖 Fixed in [7fef73929](https://github.com/Significant-Gravitas/AutoGPT/commit/7fef73929a56883a55cc23aa0bb7cd697fd03a89): added `and not ended_with_stream_error` to the re-prompt guard at service.py:3142. If a transient API error or idle timeout broke out of the inner `for` loop after `pending_thinking_only_reprompt` was set, we now fall through to `break` instead of firing `client.query()` on a broken connection.

File: autogpt_platform/backend/backend/copilot/sdk/service.py (hunk 1 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py]: 🤖 Fixed in [7fef73929](https://github.com/Significant-Gravitas/AutoGPT/commit/7fef73929a56883a55cc23aa0bb7cd697fd03a89): collapsed the 6-line narrative to a one-liner at service.py:3145.

**Review by ntindle — DISMISSED** [review]

**Comment by majdyz** [comment]
Addressed the post-refactor findings in [`fda9070200`](https://github.com/Significant-Gravitas/AutoGPT/commit/fda9070200):

| Severity | Source | Finding | Fix |
|----------|--------|---------|-----|
| 🔴 critical | coderabbit | `_consume_sdk_until_done` only flips `ended_with_stream_error`; idle/transient/breaker error metadata is dropped | Promoted `stream_error_msg` + `stream_error_code` to fields on `_SDKLoopState`; helper writes through `loop_state`; caller raises `_HandledStreamError` with `loop_state.stream_error_msg`/`code`. |
| 🟠 high | sentry | `stream_error_msg` / `stream_error_code` never updated → outer retry decision uses stale `None` | Same fix — same root cause. |
| 🟡 medium | sentry | `acc.has_tool_results` not reset on re-prompt → second-round pre-text placeholder branch fires on a stale tool_result | Reset `acc.has_tool_results = False` alongside the other re-prompt resets. |
| 🟢 CI | pyright | 5× `"ended_with_stream_error" is possibly unbound` in `stream_chat_completion_sdk` | Initialise it before the retry loop's `try:` so post-loop guards always see a bound name. Same fix unblocks the `UnboundLocalError` failures in `retry_scenarios_test.py` (test 3.11/3.12/3.13 jobs). |

Local: 48 retry-scenarios tests + 150 changed-file unit tests green; ruff + pyright clean for `service.py`.

File: autogpt_platform/backend/backend/copilot/sdk/response_adapter_test.py [hunk 5]
```diff
@@ -493,10 +486,67 @@ def test_result_success_synthesizes_fallback_text_when_final_turn_is_thinking_on
...
+        num_turns=4,
+        session_id="s1",
+        result="",
+    )
+    results = adapter.convert_message(msg)
+
+    text_deltas = [r for r in results if isinstance(r, StreamTextDelta)]
+    assert len(text_deltas) == 1
+    assert text_deltas[0].delta == "(Done — no further commentary.)"
     assert isinstance(results[-1], StreamFinish)
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/response_adapter_test.py:688]: Done in [99b0aff4d2](https://github.com/Significant-Gravitas/AutoGPT/commit/99b0aff4d2): added `test_tool_result_clears_stale_thinking_so_fallback_does_not_leak_pre_tool_thinking` which exercises the full pre-tool ThinkingBlock → ToolUseBlock → ToolResultBlock → silent finish flow, asserting the placeholder fires (not the stale pre-tool reasoning).

File: autogpt_platform/backend/backend/copilot/sdk/response_adapter.py [hunk 6]
```diff
@@ -437,10 +469,18 @@ def convert_message(self, sdk_message: Message) -> list[StreamBaseResponse]:
...
                 self._end_reasoning_if_open(responses)
                 self._ensure_text_started(responses)
+                # Prefer promoting the most recent thinking block to visible
+                # text — when the model packed the actual answer into
+                # extended-thinking instead of emitting a TextBlock, the
+                # placeholder hides a real response that's already in hand.
+                fallback_text = (
+                    self._last_thinking_content.strip()
+                    or "(Done — no further commentary.)"
+                )
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/response_adapter.py:484]: Done in [99b0aff4d2](https://github.com/Significant-Gravitas/AutoGPT/commit/99b0aff4d2): `_last_thinking_content = ""` is now cleared in the tool_result branch alongside `_text_since_last_tool_result` and `_any_tool_results_seen` (response_adapter.py around line 384).

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 7]
```diff
@@ -2525,542 +2534,567 @@ async def _run_stream_attempt(
...
+                acc.stream_completed = False
+                logger.info(
+                    "%s Re-prompting model for closing summary "
+                    "after thinking-only final turn",
+                    ctx.log_prefix,
+                )
+                await client.query(
+                    _THINKING_ONLY_REPROMPT,
+                    session_id=ctx.session_id,
+                )
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py]: Done in [99b0aff4d2](https://github.com/Significant-Gravitas/AutoGPT/commit/99b0aff4d2): the `thinking_only_reprompted` cap now lives on `_RetryState` (per-turn, persists across attempts) and is propagated to a freshly-constructed `SDKResponseAdapter` on the retry path so a transient API retry mid-turn cannot unlock another re-prompt round.

[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py]: Done in [99b0aff4d2](https://github.com/Significant-Gravitas/AutoGPT/commit/99b0aff4d2): added `_strip_synthetic_reprompt_from_cli_jsonl` and wired it into the upload path before `upload_transcript()`. The persisted JSONL drops user lines whose text equals `_THINKING_ONLY_REPROMPT`, so `--resume` on the next turn no longer replays it as a phantom user message. Coverage in `service_test.TestStripSyntheticReprompt` for list/string content, image-block preservation, empty input, malformed lines.

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 8]
```diff
@@ -2525,542 +2595,569 @@ async def _run_stream_attempt(
...
+                state.adapter.thinking_only_reprompted = True
+                # Re-prompt round must still trip the placeholder guard if model returns thinking-only again.
+                state.adapter._text_since_last_tool_result = False
+                acc.stream_completed = False
+                logger.info(
+                    "%s Re-prompting model for closing summary "
+                    "after thinking-only final turn",
+                    ctx.log_prefix,
+                )
+                await client.query(
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py]: Done in [fda9070200](https://github.com/Significant-Gravitas/AutoGPT/commit/fda9070200): `acc.has_tool_results = False` is now reset alongside the other re-prompt resets (service.py around line 3192) so the second-round pre-text placeholder branch does not fire on the prior round's tool_result.

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 9]
```diff
@@ -3118,7 +3246,7 @@ async def _run_stream_attempt(
     # already_yielded=False for transient_api_error: StreamError was NOT
     # sent to the client yet (the outer loop does it when retries are
     # exhausted, avoiding a premature error flash before the retry).
-    if ended_with_stream_error:
+    if loop_state.ended_with_stream_error:
         raise _HandledStreamError(
             "Stream error handled",
             error_msg=stream_error_msg,
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py]: Done in [fda9070200](https://github.com/Significant-Gravitas/AutoGPT/commit/fda9070200): promoted `stream_error_msg` and `stream_error_code` to fields on `_SDKLoopState`. The helper's idle/transient/breaker branches now write through `loop_state.stream_error_msg/code`, and the caller raises `_HandledStreamError` with those values so the outer retry loop sees the right code (transient_api_error / idle_timeout / etc.) instead of generic `None`.

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 10]
```diff
@@ -191,6 +191,641 @@ class _SystemPromptPreset(SystemPromptPreset, total=False):
...
+    plain module-level async generator with a small fixed parameter list,
+    instead of nesting the entire loop body deep inside
+    ``_run_stream_attempt`` under a ``while True:`` wrapper.
+    """
+
+    last_real_msg_time: float
+    last_flush_time: float
+    msgs_since_flush: int = 0
+    consecutive_empty_tool_calls: int = 0
+    ended_with_stream_error: bool = False
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py:256]: Done in [fda9070200](https://github.com/Significant-Gravitas/AutoGPT/commit/fda9070200): same fix as the sentry HIGH thread — `stream_error_msg` and `stream_error_code` are now carried on `_SDKLoopState` and threaded through `_HandledStreamError`.

**Comment by majdyz** [comment]
**On the `codecov/patch/platform-backend` flag (67.09% vs 80% target):**

The patch coverage drop is an artifact of the helper extraction in [`71c65a7ce8`](https://github.com/Significant-Gravitas/AutoGPT/commit/71c65a7ce8), not a coverage regression on new logic.

What changed in that commit:
- Moved the body of `async for sdk_msg in _iter_sdk_messages(client):` (~535 lines) out of `_run_stream_attempt` and into the new module-level helper `_consume_sdk_until_done`.
- Renamed five local variables to fields on `_SDKLoopState` (`_last_real_msg_time → loop_state.last_real_msg_time`, etc.).
- Indent-shifted every line in the body by -8 spaces.

Codecov diffs every line whose text changed — including pure indent + rename moves — and counts them as uncovered patch lines unless tests reach them. Those lines are the same lines that were untested in the *previous* commit too; they were never directly exercised by `retry_scenarios_test` (which mocks at the transcript-helper level rather than driving `_consume_sdk_until_done`'s heartbeat / idle / flush branches).

**What IS covered:**
- `_strip_synthetic_reprompt_from_cli_jsonl` — 6 dedicated tests (`TestStripSyntheticReprompt`).
- `_extract_user_message_text` — exercised through the strip helper.
- Adapter's `pending_thinking_only_reprompt` / `_last_thinking_content` reset / promote-thinking fallback — 4 tests.
- The cross-tool-boundary regression coderabbit asked for — `test_tool_result_clears_stale_thinking_so_fallback_does_not_leak_pre_tool_thinking`.
- All retry-scenario flows (48 passing).
- All other CI checks: type-check (3.11/3.12/3.13) ✅, test (3.11/3.12/3.13) ✅, lint ✅, e2e ✅, scope ✅, conflicts ✅.

**What's uncovered:**
- The moved body of `_consume_sdk_until_done` — same coverage status as before the refactor. Adding direct integration tests for that helper requires standing up `_StreamContext`/`_RetryState`/`_StreamAccumulator` mocks that don't exist today; happy to do as a follow-up.

The substantive bot findings (sentry HIGH, coderabbit CRITICAL on error-detail propagation, coverage on stale thinking, has_tool_results reset, --resume JSONL leak, per-turn cap on retries) are all addressed in the four follow-up commits, with coderabbit having confirmed each one. ntindle approved earlier in the cycle.

Requesting merge with the codecov flag accepted as informational — happy to file a follow-up ticket for direct `_consume_sdk_until_done` coverage if that's preferred over override.

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 11]
```diff
@@ -2523,544 +3160,43 @@ async def _run_stream_attempt(
...
-                        flush_err,
-                    )
-                _last_flush_time = now
-                _msgs_since_flush = 0
-
-            if acc.stream_completed:
-                break
+                yield ev
     finally:
         await _safe_close_sdk_client(sdk_client, ctx.log_prefix)
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py:3325]: Done in [`26cf25de80`](https://github.com/Significant-Gravitas/AutoGPT/commit/26cf25de80): `_strip_synthetic_reprompt_from_cli_jsonl` now drops the empty thinking-only AssistantMessage that immediately preceded the synthetic user line. Without that, post-strip JSONL went `assistant (empty) → assistant (real reply)` — orphan assistant turn that breaks Anthropic's resume role-alternation. New tests `test_also_drops_preceding_empty_thinking_only_assistant` and `test_keeps_preceding_assistant_with_real_text` lock both branches in.

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 12]
```diff
@@ -2523,544 +3206,43 @@ async def _run_stream_attempt(
...
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py:3325]: Done in [`9f2e77fd69`](https://github.com/Significant-Gravitas/AutoGPT/commit/9f2e77fd69): `loop_state.consecutive_empty_tool_calls = 0` reset alongside the other re-prompt resets (text-since-last-tool, has_tool_results) so a borderline round-1 streak doesn't trip the breaker prematurely on the first re-prompt AssistantMessage.

File: autogpt_platform/backend/backend/copilot/sdk/response_adapter.py [hunk 13]
```diff
@@ -437,10 +474,18 @@ def convert_message(self, sdk_message: Message) -> list[StreamBaseResponse]:
...
                 responses.append(
                     StreamTextDelta(
                         id=self.text_block_id,
-                        delta="(Done — no further commentary.)",
+                        delta=fallback_text,
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/response_adapter.py:488]: 🤖 Sentry's auto-comment already cites the fix at [`300bad2`](https://github.com/Significant-Gravitas/AutoGPT/commit/300bad297be8cb3ea5d9bde85c07bd4816123e24) — promotion of `_last_thinking_content` to a user message now goes through the consume helper with the additional roundtrip coverage. Marking resolved.

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 14]
```diff
@@ -2523,544 +3206,47 @@ async def _run_stream_attempt(
...
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py:3325]: Done in [`e3ab0188dd`](https://github.com/Significant-Gravitas/AutoGPT/commit/e3ab0188dd): `loop_state.last_real_msg_time = time.monotonic()` reset alongside the other re-prompt resets (text-since-last-tool, has_tool_results, consecutive_empty_tool_calls). Each round now gets its own independent idle-timeout window, so a long round 1 cannot bleed into the re-prompt's clock and trip a phantom 30-min idle abort.

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 15]
```diff
@@ -239,6 +239,695 @@ async def _resolve_dynamic_max_budget_usd(user_id: str | None) -> float:
...
+    if not isinstance(content, list):
+        return False
+    for block in content:
+        if not isinstance(block, dict):
+            return False
+        btype = block.get("type")
+        if btype == "thinking":
+            continue  # thinking-only counts as empty for our purposes
+        if btype == "text" and not (block.get("text") or "").strip():
+            continue
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py:910]: Done in [`335f35d036`](https://github.com/Significant-Gravitas/AutoGPT/commit/335f35d036): `_is_empty_assistant_entry` now treats both `thinking` and `redacted_thinking` blocks as empty for role-alternation purposes. New test `test_drops_preceding_redacted_thinking_only_assistant` locks the behaviour in.

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 16]
```diff
@@ -2571,544 +3257,52 @@ async def _run_stream_attempt(
...
-                        "%s Intermediate flush failed: %s",
-                        ctx.log_prefix,
-                        flush_err,
-                    )
-                _last_flush_time = now
-                _msgs_since_flush = 0
-
-            if acc.stream_completed:
-                break
+                yield ev
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py:3322]: Done in [`9195c23c92`](https://github.com/Significant-Gravitas/AutoGPT/commit/9195c23c92): `acc.has_appended_assistant = False` reset alongside the other re-prompt state resets so the dispatch loop allocates a fresh `ChatMessage` for the re-prompt's first text delta instead of fusing it into the empty thinking-only row.

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 17]
```diff
@@ -2571,544 +3257,57 @@ async def _run_stream_attempt(
...
     finally:
         await _safe_close_sdk_client(sdk_client, ctx.log_prefix)
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py:3325]: Done in [`03fdcc9a40`](https://github.com/Significant-Gravitas/AutoGPT/commit/03fdcc9a40): allocate a fresh `ChatMessage(role='assistant', content='')` and clear `accumulated_tool_calls` in the re-prompt block alongside the other resets. Round 2 now starts with a clean accumulator — no stale `tool_calls` to fuse with the new text.

File: autogpt_platform/backend/backend/copilot/sdk/service.py [hunk 18]
```diff
@@ -2571,544 +3257,64 @@ async def _run_stream_attempt(
...
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/sdk/service.py:3325]: Done in [`11a75a8d21`](https://github.com/Significant-Gravitas/AutoGPT/commit/11a75a8d21): `state.adapter._last_thinking_content = ""` reset alongside the other re-prompt state resets so round 2 either surfaces its own fresh thinking content or falls through to the placeholder, never round 1's stale reasoning.

**Comment by majdyz** [comment]
## /pr-test results — post-merge dev validation

Run against `https://dev-builder.agpt.co` on `2026-05-05` after the dev deploy at 02:17 UTC. Login: `<dev-login>` (credentials elided).

Companion branch with all artefacts: [`test-screenshots/pr-12992-13002`](https://github.com/Significant-Gravitas/AutoGPT/tree/test-screenshots/pr-12992-13002).

### Test 1 — Re-prompt golden path (issue 5) — ✅ PASS

> "What are the best restaurants in London? Use web search and give a comprehensive list with at least 8 entries grouped by neighborhood."

Session: `1a72e9ba-583a-4b5c-9866-685ad17bc0ec`. Footer: **"Thought for 2m 16s"** — extended thinking active, exactly the condition that produced the original `(Done — no further commentary.)` placeholder. Got a 4347-char structured list (Covent Garden / Mayfair / Shoreditch / Soho / Bethnal Green / Kensington / etc.). **No placeholder appeared.**

The Langfuse trace metadata for this turn does NOT carry `thinking_only_reprompted: true` — the model emitted a TextBlock directly without the re-prompt fallback needing to fire. The fallback chain (re-prompt → promote-thinking → placeholder) is in the deployed code and would activate if the thinking-only condition recurs.

Test 1 — restaurants

### Test 2 — Prisma fix (issue 2) — ✅ PASS

Session: `1380e1d6-2491-4354-88b0-f7da0ce17ff0`. Two callers exercise the same `manager.write_file → workspace_db().get_workspace_total_size()` Prisma codepath:

- **`AIImageGeneratorBlock`** — the headline-failing block from the original Discord report. Image rendered inline.
- **`write_workspace_file`** copilot tool — `notes.md` saved (11 bytes).

`gcloud logging read` against `dev-agpt` namespace for `ClientNotConnectedError` since deploy at `2026-05-05T02:18Z`: **zero matches.**

Test 2 — image gen + workspace file

### Test 3 — Multi-tool reasoning regression — ✅ PASS

> "Find the top-5 starred Rust repos on GitHub and summarise each in one paragraph."

Same session (follow-up). Multiple web searches + a coherent final summary covering deno, tauri, etc. No regression from the helper extraction (`_consume_sdk_until_done`).

Test 3 — multi-tool

### Test 6 — Plain Q&A regression — ✅ PASS

> "Hello, how are you today?"

Welcome message returned in ~10s. No regression.

Test 6 — plain Q&A

### Test 7 — Refresh / `--resume` regression — ✅ PASS

Reloaded the 4358-char Test 1 chat. History restored cleanly. No role-alternation error, no 500 on session GET. Confirms `_strip_synthetic_reprompt_from_cli_jsonl` + `_is_empty_assistant_entry` (including `redacted_thinking` handling) work correctly for resume.

Test 7 — refresh

### Verdict

**SAFE IN DEV** — both headline failures (`(Done — no further commentary.)` placeholder + `ClientNotConnectedError` on workspace writes) are resolved end-to-end. No hotfix needed.

🤖 1 image link(s) in this comment point at files that were already missing from the `test-screenshots/*` branch before this rewrite; they were broken and stay broken. The rest now render from GitHub attachments.
